src/_pytest/scope.py

```diff
@@ -79,6 +79,9 @@ def from_user(
                 ),
                 pytrace=False,
             )
+            # I don't know why mypy complains here, `fail()` is typed as `typing.NoReturn`
+            # yet it complains of error: Missing return statement [return] ?
+            assert False, "unreachable"
```

Comment:
Anyone savvy with mypy knows why it complains of a missing return here?  the `Scope` instantiation is correctly catching the ValueError and `fail()` has a forward ref return type of `"No Return"`.  This seems _unnecessary?_ to me

Reply:
cc @bluetech

Reply:
Looks like a mypy bug or at least something mypy can improve. Maybe the "missing return value" analysis does not consider unreachability (which *is* correctly detected in this case, I verified this by adding some statement after the fail and seeing that mypy complains that the statement is unreachable).

If you want to avoid the assert you can replace `return ...` with `scope = ...` and `return scope` outside of the try-except. Slightly less clean but OK I think.

Reply:
Whatever you folks think is best I will happily implement, kind of a rock and a hard place here :)

 - keep assert as is
 - hold a reference `scope` and return it after the except

(may have some issues around local variables referenced before assignment that won't break anything in the latter)

In the mean time I will try recreate and open a mypy issue (if one is required)

Reply:
I'd go with "hold a reference scope and return it after the except".

Reported to mypy would be nice, if you are able to reduce it to a standalone example. You should check if it has already been reported though.

Reply:
@bluetech Will do! thanks